snuba/query/expressions.py

```diff
@@ -42,6 +44,46 @@ def __iter__(self) -> Iterator[Expression]:
         """
         raise NotImplementedError
 
+    @abstractmethod
+    def accept(self, visitor: ExpressionVisitor[TVisited]) -> TVisited:
+        """
+        Accepts a visitor class to traverse the tree. The only role of this method is to
+        call the right visit method on the visitor object. Requiring the implementation
+        to call the method with the right type forces us to keep the visitor interface
+        up to date every time we create a new subclass of Expression.
+        """
+        raise NotImplementedError
+
+
+class ExpressionVisitor(ABC, Generic[TVisited]):
```

Comment:
For reference: https://en.wikipedia.org/wiki/Visitor_pattern

I looked into two possible designs for this:
1) The visitor pattern, which is commonly used to traverse and evaluate ASTs. The advantage is that we are sure the Expression class hierarchy cannot evolve without providing the needed changes to the formatter. 
2) Building a parallel hierarchy of Formatters in Clickhouse land. While that has some nice advantages (like being able to have a subclass of FunctionCall for Conditions only during formatting) it would be a lot more verbose and there would be no real way to ensure the two hierarchies evolve consistently.

Thus I went for the Visitor